### ISSUE TITLE
subsetting data in get_*() functions

### ISSUE THREAD

**Issue by jimcasaer** [inbo/camtrapdp#13]
@damianooldoni 

would it not be more logical to start all get_functions based on result of read_camtrap_dp() ? 
e.g.
data <- read_camtrap_dp(......)
get_dep_no_obs(data)
get_n_obs(data)
.....

the package knows that data contains a data$deployments and data$observations ?

**Comment by damianooldoni** [comment]
Yes, I was focusing on visualization details in previews work, but this is indeed true. It's better to pass the camtrap object (a list) to all functions. Easier for user and also more logical.

**Comment by peterdesmet** [comment]
Exactly, see also my comment on #13.

**Comment by jimcasaer** [comment]
creates however a new problem / question 
how to subset observations for species / specific tags / period is you start from the list ?

- [ ] allow selection in $observations through the get-*function
- [ ] allow users to specify specific deployments / observations (both as dataframes - rather than specifying the datapackage ?

**Comment by damianooldoni** [comment]
I think this has been partly tackled by allowing the get_* functions to select some species. An idea could be to filter on deployments too. Something like this?

```r
# using deployment ID
`get_n_obs(camtrapdp, species = "rattus norvegicus", deployment = "3b94f935-020f-49d8-bd59-1736f8a266c4")`
# using location ID 
`get_n_obs(camtrapdp, species = "rattus norvegicus", deployment = "9a72bd3a-2243-4ced-af9f-34df32483e55")`
# using location name
`get_n_obs(camtrapdp, species = "rattus norvegicus", deployment = "B_ML_val 05_molenkreek")`
```

Obviously same argument can be added to `get_species()`, `get_rai()`.

**Comment by jimcasaer** [comment]
@damianooldoni for selecting deployments users will rather use the time (year(s), month(s)), a geographic part of the study area and /or tags used in the deployment description

**Comment by damianooldoni** [comment]
I think the best solution to account all possible filtering is to use **predicates**. Something I learned from `rgbif` package v3.x: https://docs.ropensci.org/rgbif/reference/download_predicate_dsl.html

A search predicate wants to arguments: the **argument** to filter on and its **value**:

```r
basic predicate : arg = value
pred(arg = "deployment", value = "9a72bd3a-2243-4ced-af9f-34df32483e55")
# get n observations for deployment: "9a72bd3a-2243-4ced-af9f-34df32483e55"
get_n_obs(camtrapdp, pred(arg = "deployment", value = "9a72bd3a-2243-4ced-af9f-34df32483e55"))
```

Translated to our data:

```r
#' get number of observations of A. platyrhynchos and start deployment > 2020
get_n_obs(camtrapdp$observations, pred("scientific_name", "Anas platyrhynchos"),pred("start", "2020"))

# same result using pred_and()
get_n_obs(camtrapdp$observations, pred_and(pred("scientific_name", "Anas platyrhynchos"),pred("start", "2020")))

#' you can go as far as you want in making complex queries
#' get number of observations of A. platyrhynchos AND (start deployment > 2020 OR location_name one of 
#' "B_ML_val 05_molenkreek", "B_ML_val 06_Oostpolderkreek")
get_n_obs(camtrapdp$observations,
    pred_and(pred("scientific_name", "Anas platyrhynchos"),
             pred_or(pred("start", "2020"),
                     pred_in("location_name", c("B_ML_val 05_molenkreek", "B_ML_val 06_Oostpolderkreek")))))
```

This solution is slightly less readable than adding arguments as the columns we like to use for the data filtering, but allows us to use **all kind of logical operators** (and, `&`, or, `|`) and all their combinations, plus all mathematical operators: `pred_gt` (greater than), `pred_lt` (less than), `pred_notna` (not NA), `pred_in` (one of)! Not only, it makes the maintenance of the functions easier as we don't have to change anything if the cameratrap data standard slightly changes by adding new columns or changing some column names.
 
The only difference with rgbif is that rgbif's predicates translate your predicates to JSON syntax, while in our case they will build a `filter()` statement and therefore they could be programmed quite easily without too much higher structures.

@jimcasaer: What do you think?

**Comment by jimcasaer** [comment]
sounds logic to me ! @damianooldoni

**Comment by damianooldoni** [comment]
Solved by #29
